cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1160 @@
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+)
+
+// PreTableInfo returns the table info which will be overwritten by the specified job
+func (s *Snapshot) PreTableInfo(job *timodel.Job) (*model.TableInfo, error) {
+	switch job.Type {
+	case timodel.ActionCreateSchema, timodel.ActionModifySchemaCharsetAndCollate, timodel.ActionDropSchema:
+		return nil, nil
+	case timodel.ActionCreateTable, timodel.ActionCreateView, timodel.ActionRecoverTable:
+		// no pre table info
+		return nil, nil
+	case timodel.ActionRenameTable, timodel.ActionDropTable, timodel.ActionDropView, timodel.ActionTruncateTable:
+		// get the table will be dropped
+		table, ok := s.PhysicalTableByID(job.TableID)
+		if !ok {
+			return nil, cerror.ErrSchemaStorageTableMiss.GenWithStackByArgs(job.TableID)
+		}
+		return table, nil
+	case timodel.ActionRenameTables:
+		// DDL on multiple tables, ignore pre table info
+		return nil, nil
+	default:
+		binlogInfo := job.BinlogInfo
+		if binlogInfo == nil {
+			log.Warn("ignore a invalid DDL job", zap.Reflect("job", job))
+			return nil, nil
+		}
+		tbInfo := binlogInfo.TableInfo
+		if tbInfo == nil {
+			log.Warn("ignore a invalid DDL job", zap.Reflect("job", job))
+			return nil, nil
+		}
+		tableID := tbInfo.ID
+		table, ok := s.PhysicalTableByID(tableID)
+		if !ok {
+			return nil, cerror.ErrSchemaStorageTableMiss.GenWithStackByArgs(job.TableID)
+		}
+		return table, nil
+	}
+}
+
+// NewSingleSnapshotFromMeta creates a new single schema snapshot from a tidb meta
+func NewSingleSnapshotFromMeta(meta *timeta.Meta, currentTs uint64, forceReplicate bool) (*Snapshot, error) {
+	// meta is nil only in unit tests
+	if meta == nil {
+		snap := NewEmptySnapshot(forceReplicate)
+		snap.currentTs = currentTs
+		return snap, nil
+	}
+	return NewSnapshotFromMeta(meta, currentTs, forceReplicate)
+}
+
+// Snapshot stores the source TiDB all schema information and it should be READ-ONLY!
+// If no special comments, all public methods are thread-safe.
+type Snapshot struct {
+	// map[versionedEntityName] -> int64
+	// The ID can be `-1` which means the table is deleted.
+	tableNameToID *btree.BTree
+
+	// map[versionedEntityName] -> int64
+	// The ID can be `-1` which means the table is deleted.
+	schemaNameToID *btree.BTree
+
+	// map[versionedID] -> *timodel.DBInfo
+	// The target can be `nil` which means the entity is deleted.
+	schemas *btree.BTree
+
+	// map[versionedID] -> *model.TableInfo
+	// The target can be `nil` which means the entity is deleted.
+	tables *btree.BTree
+
+	// map[versionedID] -> *model.TableInfo
+	partitions *btree.BTree
+
+	// map[versionedID] -> struct{}
+	truncatedTables *btree.BTree
+
+	// map[versionedID] -> struct{}
+	// Partitions and tables share ineligibleTables because their IDs won't conflict.
+	ineligibleTables *btree.BTree
+
+	// if forceReplicate is true, treat ineligible tables as eligible.
+	forceReplicate bool
+
+	currentTs uint64
+
+	rwlock *sync.RWMutex
+	locked int
```

Comment:
Can we remove `locked`?

Reply:
`locked` forbids a thread to try lock multiple times.

Reply:
You are right. The design is wrong.

Reply:
Recursive locking is quite tricky in Go, because there is no "goroutine-local" storage.